Improve implementation of Mbed TLS timing

LDong-Arm opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

This PR makes the following fixes and improvements to the Mbed OS implementation of Mbed TLS timing:
* Switches to use `LowPowerTimer` or `Timer` for time measurements. The Mbed OS implementation of [gettimeofday()](https://github.com/ARMmbed/mbed-os/blob/0c9be22f58a75c2dee5902525b953aec21e0ace5/platform/source/mbed_rtc_time.cpp#L95-L115) is only precise to seconds, but Mbed TLS timing requires milliseconds.
* Uses `LowPowerTimeout` (instead of `Timeout`) for mbedtls_set_alarm() if available, to save power.
* Adds a full implementation of `timing.h`, to fix "undefined symbol" errors generated by the Arm Compiler. This will also benefits user applications if they need to enable timing in Mbed TLS.
* Adds `mbedtls_timing_self_test` to the existing Greentea test `connectivity-mbedtls-tests-TESTS-mbedtls-selftest`, if `MBEDTLS_TIMING_C` is available.
* Unsets MBEDTLS_SELF_TEST from the default configuration, and adds `TESTS/configs/mbedtls.json` to enable self tests and timing.

Please see the commit messages for reasoning.

To run the timing self test (e.g. on K64F, with the ARM toolchain):
```
mbed test -t ARM -m K64F -n connectivity-mbedtls-tests-TESTS-mbedtls-selftest --app-config TESTS/configs/mbedtls.json
```

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

Now compilation failure with the Arm Compiler with `MBEDTLS_TIMING_C` is fixed. For example, the benchmark application in [mbed-os-example-tls](https://github.com/ARMmbed/mbed-os-example-tls) depends on that.

Additionally, a full implementation of Mbed TLS timing becomes available if an application needs it, and it can be enabled by defining `MBEDTLS_TIMING_C` and `MBEDTLS_TIMING_ALT`.

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

None.

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

None.

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [x] Tests / results supplied as part of this PR
    
Manually ran the self test (see description above) and got
```
test case: 'mbedtls_timing_self_test' ........................................................ OK in 2.56 sec
```
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

@ARMmbed/mbed-os-core @Patater @0xc0170 

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

LDong-Arm committed:
mbedtls: Use LowPowerTimer/Timer for timing

Previously we used `gettimeofday()` for Mbed TLS timing, but its
implementation provided by Mbed OS is only precise to seconds. The
microsecond component of the output `struct timeval` is always set
to zero. But Mbed TLS requires millisecond precision.

To provide required timing precision, switch to use `LowPowerTicker`
or (microsecond) `Ticker`. `LowPowerTicker` is preferred as it saves
power and Mbed TLS does not require microsecond precision.

LDong-Arm committed:
mbedtls: Add full platform implementation of timing

When MBEDTLS_TIMING_C and MBEDTLS_TIMING_ALT are enabled,
the Arm Compiler generates errors like the following (one for
each missing symbol):

    Error: L6218E: Undefined symbol mbedtls_timing_get_delay

Reason:

The function `mbedtls_timing_self_test()` in the Mbed TLS default
`timing.c` always gets compiled, if MBEDTLS_SELF_TEST is defined.
And MBEDTLS_SELF_TEST is always defined, as we have a Greentea test
to run some of the Mbed TLS self tests. (In the future we should try
not to enable MBEDTLS_SELF_TEST except for tests, but it requires
a rework in our test flow.)

`mbedtls_timing_self_test()` tests (calls) the full API declared in
`timing.h`, and the ARM Compiler requires all symbols referenced by
all functions to be defined, even those not used by the final
application. This is unlike GCC_ARM which resolves what are required.

Solution:

To fix the "undefined symbol" errors, we add an implementation of
`mbedtls_timing_get_timer()` based on Mbed OS `LowPowerTimer` or
`Timer` (depending on which one is available), and copy Mbed TLS's
default `mbedtls_timing_set_delay()` and `mbedtls_timing_get_delay()`
which are built on top of `mbedtls_timing_get_timer()`. This will also
benefit user applications that need to enable timing in Mbed TLS.

LDong-Arm committed:
mbedtls: Run mbedtls_timing_self_test if MBEDTLS_TIMING_C

This allows us to verify the support for Mbed TLS timing on Mbed OS.

Note: The macros MBEDTLS_TIMING_C and MBEDTLS_TIMING_ALT are not
enabled by default and need to be additionally enabled to run this
test.

Patater commented on connectivity/mbedtls/platform/inc/timing_alt.h:
```diff
@@ -24,11 +24,9 @@
 struct mbedtls_timing_hr_time
 {
-    struct timeval start;
+    unsigned long start;
```
Why not make `gettimeofday()` more precise (to ms)? Wouldn't that benefit more use cases?

LDong-Arm replied:
Discussed offline: The current implementation of `gettimeofday()` in Mbed OS is based on the RTC HAL which is precise to one second. Supporting higher precision would probably involve bringing tickers into scope.

Patater replied:
OK, good to follow up on later. If we add `alarm()` and `signal()` to Mbed OS, we may be able to use the default timing implementation from Mbed TLS and not have to implement our own, given we have a `gettimeofday()` precise within 100ms as the timing self-test requires.

Patater commented on connectivity/mbedtls/platform/src/timing_mbed.cpp:
```diff
@@ -46,22 +48,25 @@ extern "C" void mbedtls_set_alarm(int seconds)
-    gettimeofday(&tv_cur, NULL);
-    return((tv_cur.tv_sec - tv_init.tv_sec) * 1000000
-          + (tv_cur.tv_usec - tv_init.tv_usec));
+    return(timer.elapsed_time().count());
```
Style:
```diff
- return(timer.elapsed_time().count());
+ return timer.elapsed_time().count();
```

LDong-Arm replied:
Done

Patater commented on connectivity/mbedtls/platform/src/timing_mbed.cpp:
```diff
@@ -46,22 +48,25 @@ extern "C" void mbedtls_set_alarm(int seconds)
+#elif DEVICE_USTICKER
+    static mbed::Timer timer;
+#else
+#error "MBEDTLS_TIMING_C required either LPTICKER or USTICKER"
```
In commit "mbedtls: Use LowPowerTimer/Timer for timing"
```diff
- required
+ requires
```

LDong-Arm replied:
Done

Patater replied:
Please fix this in the commit "mbedtls: Use LowPowerTimer/Timer for timing"

LDong-Arm replied:
Done

Patater commented on connectivity/mbedtls/platform/src/timing_mbed.cpp:
```diff
@@ -46,27 +47,86 @@ extern "C" void mbedtls_set_alarm(int seconds)
+        return 1;
+
+    return 0;
+}
```
Missing newline at end of file

LDong-Arm replied:
Done

LDong-Arm committed:
timing_mbed.cpp: Check MBEDTLS_TIMING_ALT

Do not compile the Mbed implementation of Mbed TLS unless
MBEDTLS_TIMING_ALT is defined. This prevents a macro check error on
devices that do not have LPTICKER or USTICKER when Mbed TLS timing
is not enabled.

Patater commented on connectivity/mbedtls/platform/src/timing_mbed.cpp:
```diff
@@ -1,6 +1,7 @@
 /*
  *  timing.cpp
  *
+ *  Copyright The Mbed TLS Contributors
```
Given that the previous commit switched away from using `timeval`, does the commit message for "mbedtls: Add full platform implementation of timing" still make sense? Probably we don't have `error: field has incomplete type 'struct timeval'` anymore.

LDong-Arm replied:
Removed from the commit message

Patater commented on connectivity/mbedtls/platform/src/timing_mbed.cpp:
```diff
@@ -1,6 +1,7 @@
  *  timing.cpp
  *
+ *  Copyright The Mbed TLS Contributors
  *  Copyright (C) 2021, Arm Limited, All Rights Reserved
```
In the commit message for "mbedtls: Add full platform implementation of timing"
```diff
- And MBEDTLS_SELF_TEST is defined
+ And MBEDTLS_SELF_TEST is always defined
```

LDong-Arm replied:
Done

LDong-Arm committed:
mbedtls: Use LowPowerTimeout for mbedtls_set_alarm() if available

The function `mbedtls_set_alarm()` is only precise to seconds, so
`LowPowerTimeout` is enough and saves power.

Patater commented on connectivity/mbedtls/tests/TESTS/mbedtls/selftest/main.cpp:
```diff
@@ -65,6 +66,10 @@ MBEDTLS_SELF_TEST_TEST_CASE(mbedtls_sha512_self_test)
 MBEDTLS_SELF_TEST_TEST_CASE(mbedtls_entropy_self_test)
 #endif
 
+#if defined(MBEDTLS_TIMING_C)
```
We need a test config json to enable TIMING_C and TIMING_ALT and SELF_TEST. This is adding dead code otherwise. We also need to update CI to use the test config json for `mbedtls/selftest`. We can also take the opportunity to disable SELF_TEST by default.

LDong-Arm replied:
I've added a json file for self tests and disabled MBEDTLS_SELF_TEST by default.

LDong-Arm committed:
Move Mbed TLS self tests to a separate configuration

We potentially save flash space by not enabling Mbed TLS self-tests
by default. A new test config file, TESTS/configs/mbedtls.json, is
provided to enable self tests. This newly created JSON file also
enables timing in Mbed TLS so timing gets tested.

LDong-Arm reviewed (commented):
@Patater Thanks for the review, I've addressed your comments.

Patater commented on connectivity/mbedtls/platform/src/timing_mbed.cpp:
```diff
@@ -46,22 +48,25 @@ extern "C" void mbedtls_set_alarm(int seconds)
 #if !defined(HAVE_HARDCLOCK)
 #define HAVE_HARDCLOCK
-#include "platform/mbed_rtc_time.h"
 static int hardclock_init = 0;
```
You might also want to rename hardclock_init in "mbedtls: Use LowPowerTimer/Timer for timing" rather than in "mbedtls: Add full platform implementation of timing".

LDong-Arm replied:
Done

Patater commented on connectivity/mbedtls/platform/src/timing_mbed.cpp:
```diff
@@ -46,27 +47,86 @@ extern "C" void mbedtls_set_alarm(int seconds)
+
+extern "C" unsigned long mbedtls_timing_get_timer( struct mbedtls_timing_hr_time *val, int reset )
+{
+    if(timer_init == 0)
```
```diff
- if(timer_init == 0)
+ if (timer_init == 0)
```

Patater commented on connectivity/mbedtls/platform/src/timing_mbed.cpp:
```diff
@@ -46,27 +47,86 @@ extern "C" void mbedtls_set_alarm(int seconds)
 }
 #endif /* !HAVE_HARDCLOCK */
+
+extern "C" unsigned long mbedtls_timing_get_timer( struct mbedtls_timing_hr_time *val, int reset )
```
For functions we implement, we should follow Mbed OS style.

```diff
- extern "C" unsigned long mbedtls_timing_get_timer( struct mbedtls_timing_hr_time *val, int reset )
+ extern "C" unsigned long mbedtls_timing_get_timer(struct mbedtls_timing_hr_time *val, int reset)
```

Patater commented on connectivity/mbedtls/platform/src/timing_mbed.cpp:
```diff
@@ -46,27 +47,86 @@ extern "C" void mbedtls_set_alarm(int seconds)
+ * Mbed OS, so we copy them over.
+ */
+
+extern "C" void mbedtls_timing_set_delay(void *data, uint32_t int_ms, uint32_t fin_ms)
```
The weird style for these functions is OK, because they are copied from Mbed TLS and we want to minimize differences for easy comparison and update.

LDong-Arm replied:
I've just run astyle to automatically fix the style of this file.

Patater commented on connectivity/mbedtls/platform/src/timing_mbed.cpp:
```diff
@@ -24,6 +24,9 @@
 #include MBEDTLS_CONFIG_FILE
 #endif
+
+#if defined(MBEDTLS_TIMING_ALT)
```
In commit message for "timing_mbed.cpp: Check MBEDTLS_TIMING_ALT".

```diff
-  This avoid macro check error
+  This avoids a macro check error
```

LDong-Arm replied:
Done

Patater commented on connectivity/mbedtls/include/mbedtls/config.h:
```diff
@@ -1396,7 +1396,7 @@
  * Enable the checkup functions (*_self_test).
  */
-#define MBEDTLS_SELF_TEST
+//#define MBEDTLS_SELF_TEST
```
In commit "Move Mbed TLS self tests to a separate configuration", please update `connectivity/mbedtls/include/mbedtls/config-no-entropy.h` too.

LDong-Arm replied:
Done

Patater commented on connectivity/mbedtls/tools/importer/adjust-config.sh:
```diff
@@ -117,6 +117,9 @@ conf unset MBEDTLS_SSL_TRUNCATED_HMAC
 
 conf unset MBEDTLS_PLATFORM_TIME_TYPE_MACRO
 
+# not needed for production code
```
Some production code might use self-tests to ensure the hardware is working correctly before continuing with other functional stages of the application. For example, your PC's BIOS likely does some self-tests before continuing with the boot. We should put the reason as something more like "potentially save flash space by not enabling self-tests by default".

LDong-Arm replied:
Done, sounds good. It looks like the linker does filter out most unused functions, but some static variables are still linked to the executable. Also seem this happening to BLE.

Patater commented on connectivity/mbedtls/tools/importer/adjust-config.sh:
```diff
@@ -117,6 +117,9 @@ conf unset MBEDTLS_SSL_TRUNCATED_HMAC
 conf unset MBEDTLS_PLATFORM_TIME_TYPE_MACRO
 
+# not needed for production code
+conf unset MBEDTLS_SELF_TEST
```
Please update `adjust-no-entropy-config.sh`, too, as the base configuration file we use for no-entropy appears to have self-tests enabled by default (https://github.com/ARMmbed/mbedtls/blob/development/configs/config-no-entropy.h#L50).

LDong-Arm replied:
Done

Patater commented on connectivity/mbedtls/tools/importer/adjust-config.sh:
```diff
@@ -117,6 +117,9 @@ conf unset MBEDTLS_SSL_TRUNCATED_HMAC
 
 conf unset MBEDTLS_PLATFORM_TIME_TYPE_MACRO
 
+# potentially save flash space by not enabling self-tests by default
```
The commit message for "Move Mbed TLS self tests to a separate configuration" should be updated, too, not just this comment. It still says "there is no point..."

LDong-Arm replied:
Done

Patater reviewed (approved):
LGTM

jeromecoutant commented on TESTS/configs/mbedtls.json:
```diff
@@ -0,0 +1,7 @@
+    "macros": [
+        "MBEDTLS_SELF_TEST",
+        "MBEDTLS_TIMING_C",
+        "MBEDTLS_TIMING_ALT"
```
_ALT should be enabled only for HW crypto targets ?

LDong-Arm replied:
Mbed TLS has a number of _ALT macros, but `MBEDTLS_TIMING_ALT` is only to enable timing based on Mbed OS API (i.e. LowPowerTicker/Ticker). This one is not related to crypto capabilities.
